In [ ]:
from tensorflow.keras.datasets import mnist

In [ ]:
(X_train, y_train), (X_test, y_test) = mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


In [ ]:
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

In [ ]:
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

In [ ]:
from tensorflow.keras.utils import to_categorical
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [ ]:
from tensorflow.keras.models import Sequential
cnn = Sequential()

In [ ]:
from tensorflow.keras.layers import Conv2D, Dense, Flatten, MaxPooling2D

In [ ]:
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))

In [ ]:
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))

In [ ]:
cnn.add(Flatten())
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

In [ ]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 114s 134ms/step - loss: 0.1382 - accuracy: 0.9577 - val_loss: 0.0411 - val_accuracy: 0.9872
Epoch 2/5
844/844 [==============================] - 115s 136ms/step - loss: 0.0419 - accuracy: 0.9870 - val_loss: 0.0304 - val_accuracy: 0.9910
Epoch 3/5
844/844 [==============================] - 112s 133ms/step - loss: 0.0273 - accuracy: 0.9916 - val_loss: 0.0383 - val_accuracy: 0.9895
Epoch 4/5
844/844 [==============================] - 113s 134ms/step - loss: 0.0188 - accuracy: 0.9944 - val_loss: 0.0459 - val_accuracy: 0.9865
Epoch 5/5
844/844 [==============================] - 112s 133ms/step - loss: 0.0148 - accuracy: 0.9951 - val_loss: 0.0364 - val_accuracy: 0.9900


Training time 10m 22s.

In [ ]:
cnn.evaluate(X_test, y_test)

313/313 [==============================] - 6s 19ms/step - loss: 0.0313 - accuracy: 0.9902


[0.0313175693154335, 0.9901999831199646]

Loss 0.0313 Accuracy 0.9902

In [ ]:
from tensorflow.keras.datasets import fashion_mnist

In [ ]:
(X_train, y_train), (X_test, y_test) = fashion_mnist.load_data()

4431872/4422102 [==============================] - 0s 0us/step


In [ ]:
X_train = X_train.reshape((60000, 28, 28, 1))
X_test = X_test.reshape((10000, 28, 28, 1))

In [ ]:
X_train = X_train.astype('float32') / 255
X_test = X_test.astype('float32') / 255

In [ ]:
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [ ]:
cnn = Sequential()

In [ ]:
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Flatten())
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

In [ ]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [ ]:
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 115s 135ms/step - loss: 0.4596 - accuracy: 0.8341 - val_loss: 0.3541 - val_accuracy: 0.8735
Epoch 2/5
844/844 [==============================] - 114s 135ms/step - loss: 0.3037 - accuracy: 0.8885 - val_loss: 0.3115 - val_accuracy: 0.8855
Epoch 3/5
844/844 [==============================] - 115s 136ms/step - loss: 0.2561 - accuracy: 0.9054 - val_loss: 0.2751 - val_accuracy: 0.8985
Epoch 4/5
844/844 [==============================] - 114s 135ms/step - loss: 0.2230 - accuracy: 0.9176 - val_loss: 0.2530 - val_accuracy: 0.9078
Epoch 5/5
844/844 [==============================] - 113s 134ms/step - loss: 0.1981 - accuracy: 0.9266 - val_loss: 0.2592 - val_accuracy: 0.9067


Training time 10m 22s.

In [ ]:
cnn.evaluate(X_test, y_test)

313/313 [==============================] - 6s 19ms/step - loss: 0.2792 - accuracy: 0.9011


[0.2791972756385803, 0.9010999798774719]

Loss 0.2792 Accuracy 0.9011

The training time was the same but the regular Mnist dataset fared better than the Fashion Mnist dataset.

In [ ]:
cnn = Sequential()
cnn.add(Conv2D(filters=64, kernel_size=(3, 3), activation='relu', input_shape=(28, 28, 1)))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Conv2D(filters=128, kernel_size=(3, 3), activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Flatten())
cnn.add(Dense(4096))
cnn.add(Dense(units=128, activation='relu'))
cnn.add(Dense(units=10, activation='softmax'))

In [25]:
cnn.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

In [26]:
cnn.fit(X_train, y_train, epochs=5, batch_size=64, validation_split=0.1)

Epoch 1/5
844/844 [==============================] - 260s 308ms/step - loss: 0.4209 - accuracy: 0.8450 - val_loss: 0.3373 - val_accuracy: 0.8770
Epoch 2/5
844/844 [==============================] - 259s 307ms/step - loss: 0.2806 - accuracy: 0.8947 - val_loss: 0.2955 - val_accuracy: 0.8893
Epoch 3/5
844/844 [==============================] - 257s 305ms/step - loss: 0.2459 - accuracy: 0.9084 - val_loss: 0.2792 - val_accuracy: 0.8965
Epoch 4/5
844/844 [==============================] - 257s 305ms/step - loss: 0.2207 - accuracy: 0.9171 - val_loss: 0.2669 - val_accuracy: 0.9015
Epoch 5/5
844/844 [==============================] - 258s 305ms/step - loss: 0.2035 - accuracy: 0.9226 - val_loss: 0.2708 - val_accuracy: 0.9055


Training time 22m 22s.

In [27]:
cnn.evaluate(X_test, y_test)

313/313 [==============================] - 13s 41ms/step - loss: 0.2875 - accuracy: 0.8996


[0.28754478693008423, 0.8996000289916992]

Loss 0.2875 Accuracy 0.8996

Adding the extra Dense layer with 4096 neurons added extra training time and resulted in lower accuracy.